src/lib.rs

```diff
@@ -0,0 +1,379 @@
+    fn remove(&mut self, ptr: usize) {
+        for block in self.iter_mut() {
+            match *block {
+                Block::Entry(base_ptr, _) => {
+                    if base_ptr.get() == ptr {
+                        *block = Block::Free;
+                        return;
+                    }
+                }
+                _ => (),
+            }
+        }
+    }
+
+    /// Given an arbitrary pointer (base or inner), finds pointer info of the
+    /// associated base pointer if it exists.
+    fn find_base(&self, ptr: usize) -> Option<PtrInfo> {
+        for block in self.iter() {
+            match *block {
+                Block::Entry(base_ptr, size) => {
+                    if ptr >= base_ptr.get() && ptr < (base_ptr.get() + size as usize) {
+                        return Some(PtrInfo {
+                            ptr: base_ptr.get(),
+                            size,
+                        });
+                    }
+                }
+                _ => (),
+            }
+        }
+        None
+    }
+
+    /// Updates the size associated with a base pointer (perfomed on a realloc).
+    /// This must not be called with an inner pointer.
+    fn update(&mut self, ptr: usize, size: usize) {
+        for block in self.iter_mut() {
+            match *block {
+                Block::Entry(base_ptr, _) => {
+                    if ptr == base_ptr.get() {
+                        *block = Block::Entry(base_ptr, size);
+                        return;
+                    }
+                }
+                _ => (),
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+enum Block {
+    Free,
+    // It is UB to call the raw allocator with a ZST. All instances in the
+    // standard library use a Unique::empty() abstraction and ensure that the
+    // raw allocator is never called in such instance. Assuming that users
+    // adhere to this too, encoding the pointer as a NonZeroUsize means that the
+    // value for 0 can be used to encode the discriminant tag. This reduces the
+    // size of a `Block` from 3 machine words to 2.
+    Entry(core::num::NonZeroUsize, usize),
+}
+
+/// The metadata for each ptr is only 2 machine words which means that copying
+/// the data out of the `AllocList` is not that expensive. This is done so that
+/// the API is cleaner:
+///     * We abstract away the complexities of the underlying block structure
+///     * It is easier to work with primitives than references to primitives
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+struct PtrInfo {
+    ptr: usize,
+    size: usize,
+}
+
+static mut PTR_METADATA: Option<AllocList> = None;
+
+fn abort() {
+    unsafe { core::intrinsics::abort() };
+}
+
+/// ============================================================================
+/// Listed below are the five allocation functions currently required by custom
+/// allocators. Their signatures and symbol names are not currently typechecked
```

Comment:
The sentence starting "their signatures" doesn't quite parse.

Reply:
This comment was copied from the allocator API documentation, I admit it doesn't make a great deal of sense to me either so I've removed it.